File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
class Solution:
  def totalFruit(self, fruits: List[int]) -> int:
    basket = collections.defaultdict(lambda :0)
    res = 0
    i = 0
    for j, fruit in enumerate(fruits):
      basket[fruit] += 1
      while len(basket) > 2:
        basket[fruits[i]] -= 1
        if basket[fruits[i]] == 0:
          basket.pop(fruits[i])
        i += 1
      res = max(res, j - i + 1)
    return res

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.js
/**
 * @param {number[]} fruits
 * @return {number}
 */
var totalFruit = function(fruits) {
  const basket = new Map();
  let res = 0;
  let i = 0;
  for (let j = 0; j < fruits.length; j++) {
    if (!basket.has(fruits[j])) {
      basket.set(fruits[j], 0);
    }
    basket.set(fruits[j], basket.get(fruits[j]) + 1);
    while (basket.size > 2) {
      basket.set(fruits[i], basket.get(fruits[i]) - 1);
      if (basket.get(fruits[i]) === 0) {
        basket.delete(fruits[i]);
      }
      i++;
    }
    res = Math.max(res, j - i + 1);
  }
  return res;
  
    
};